### DISCUSSION TITLE
Add a custom viz plugin in a production server

### DISCUSSION THREAD

**Discussion by Heg0Damask** [apache/superset#20045] [Q&A / Help]
Hello 
I saw this documentation for creating a plugin https://superset.apache.org/docs/contributing/creating-viz-plugins. 
After building the plugin how can I add it in a production server ?
Thanks

**Comment by diegomedina248** [comment]
Hi,

Assuming the example from the documentation, you should install the created plugin running:
```bash
npm i -S /tmp/superset-plugin-chart-hello-world
```

Then, you'll need to add and initialize the plugin in the file: `superset-frontend/src/visualizations/presets/MainPreset.js`
The initialization/configuration is described at the bottom of the documentation.

You can either fork the repo and add the configuration, or run a script in your production deployment pipeline to make the modifications.

**Reply by EBoisseauSierra** [reply]
Could you please expand on what is concretely required in “(fork the repo and) add the configuration”?

Say I've added the code of the plugin as part of the main Superset repo, e.g. in  `./superset-frontend/plugins/plugin-chart-hello-world`. I

1. Build the package:
   ```
   cd ./superset-frontend/plugins/plugin-chart-hello-world
   npm ci && npm run build
   cd -
   ```
2. Install the plugin:
   ```
   npm i -S ./superset-frontend/plugins/plugin-chart-hello-world
   ```
4. Reference the plugin in Python as per instructions
   ```
    diff --git c/superset-frontend/src/visualizations/presets/MainPreset.js w/superset-frontend/src/visualizations/presets/MainPreset.js
    index dc3736ff1..691a102e4 100644
    --- c/superset-frontend/src/visualizations/presets/MainPreset.js
    +++ w/superset-frontend/src/visualizations/presets/MainPreset.js
    @@ -78,6 +78,7 @@ import {
       GroupByFilterPlugin,
     } from 'src/filters/components/';
     import { PivotTableChartPlugin as PivotTableChartPluginV2 } from '@superset-ui/plugin-chart-pivot-table';
    +import { SupersetHelloWorld } from 'superset-hello-world';
     import FilterBoxChartPlugin from '../FilterBox/FilterBoxChartPlugin';
     import TimeTableChartPlugin from '../TimeTable';
     
    @@ -164,6 +165,7 @@ export default class MainPreset extends Preset {
             new TimeColumnFilterPlugin().configure({ key: 'filter_timecolumn' }),
             new TimeGrainFilterPlugin().configure({ key: 'filter_timegrain' }),
             new EchartsTreeChartPlugin().configure({ key: 'tree_chart' }),
    +        new SupersetHelloWorld().configure({ key: 'ext-hello-world' }),
             ...experimentalplugins,
           ],
         });
   ```
   
I can now access the package locally (:tada:).

However, when I try to build the Python package via
```
python setup.py sdist
```
and later install this package on a server, I don't have my custom viz plugin appearing.

**Reply by villebro** [reply]
@EBoisseauSierra before you do `python setup.py sdist` you'll need to build the frontend assets. Something like this:

```bash
cd superset-frontend
npm ci  # install the pinned frontend dependencies
patch ...  # apply the diff
npm i -S <path to plugin>  # install the new plugin
npm run build  # this builds the static assets that are served by Flask
cd ..
python setup.py sdist  # build the python package
```

**Reply by arnabneogi86** [reply]
Hi,
I'm installing superset in our production using the command - 
pip install apache-superset==2.0.1 
and the running the container using 
CMD ["gunicorn", "superset.app:create_app()"]

In my local dev environment, I can create viz plugins and run the dev server. I can also build the superset-frontend code

My question is how can i move this superset-frontend to my prod server - since I'm not running PROD using git code and using superset release version and installing using pip.

It will be really helpful if you can guide.

Thanks

**Reply by santoshmallah** [reply]
Hello, did you get a solution or any idea for this?
I also wanted to do the same
It will be really helpful if you can guide me.

**Reply by arnabneogi86** [reply]
Yes I can.
1. git clone the superset project from the tag - 2.0.1 ( I prefer releases instead of current head)
2. Create your own plugin and npm link superset-frontend with your plugin. 
3. Once you are done with your development, build your plugin and run **python setup.py sdist **in 2.0.1 folder. This will create a tar file in the dist folder.
4. Take the file in your deployed server and run pip install.
Remember this will generate only frontend code and that will be deployed.


Following versions are imp here:
Node.js: Version 16

npm: Version 7

Python : 3.8

Python in pip : 3.8
